Announce `error-stack` (#628)
hashintel/hash: sites/hashdev/src/_pages/blog/4_announcing-error-stack.mdx

---
authorPhoto: blog/4_announcing-error-stack/tim.jpg

> postPhoto: blog/2_open-source/OpenSource_Cover.jpg

Reviewer: ```suggestion
postPhoto: blog/4_announcing-error-stack/ErrorStack_Cover.jpg
```

title: "Announcing error-stack"

> subtitle: "A context-aware error library with arbitrary attached user data"

Reviewer: ```suggestion
subtitle: "A context-aware error library for Rust that supports arbitrary attached user data"
```

author: "Tim Diekmann"
jobTitle: "Platform Engineer at HASH"
date: "2022-06-10"
---

> We are proud to announce a new error library for Rust: `error-stack` - a context-aware error library with arbitrary attached user data. **[View the crate on GitHub >](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack)**

Reviewer 2: I've just looked at the newest deployment and the opening paragraph looks a little strange. 

Perhaps we should put a line-break for "View the crate on Github".

Reviewer: I agree. Opened as a suggestion: https://github.com/hashintel/hash/pull/628/files#r894458618

> We are proud to announce a new error library for Rust: `error-stack` - a context-aware error library with arbitrary attached user data. **[View the crate on GitHub >](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack)**

Reviewer: ```suggestion
We are proud to announce a new error library for Rust: `error-stack` - a context-aware error library with arbitrary attached user data.

**[View the crate on GitHub >](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack)**
```

This post outlines our trials and tribulations in error-handling, why we ultimately built the crate, what architectural choices we made in doing so, and what you can expect from it.

## Our error-handling journey

> At HASH, we've been building a simulation engine — written in Rust — called [hEngine](https://hash.ai/platform/engine). It's a large code base, with a number of contributors, and its size and complexity has grown over time (check out the code on [GitHub](https://github.com/hashintel/hash)). Across some big refactors, as well as everyday feature development, we found error handling became unwieldy and ourselves struggling to keep up. In a bid to improve our debugging process and reduce maintenance overheads we decided to revisit our strategy for managing errors.

Reviewer: ```suggestion
At HASH, we've been building a simulation engine — written in Rust — called [hEngine](https://hash.ai/platform/engine). It's a large codebase, with a number of contributors, and its size and complexity has grown over time (check out the code on [GitHub](https://github.com/hashintel/hash)). Across some big refactors, as well as everyday feature development, we found error handling became unwieldy and ourselves struggling to keep up. In a bid to improve our debugging process and reduce maintenance overheads we decided to revisit our strategy for managing errors.
```

I err towards the reasoning in https://english.stackexchange.com/a/237851

> Up until recently our approach to error handling centered around using [`thiserror`](https://crates.io/crates/thiserror), a fantastic crate which abstracts away a lot of Rust's verbosity when creating errors. We used to create enumerations (generally a large, top-level one, per crate and/or major modules) to describe our different error types, and used `thiserror`'s macros to handle a lot of the trait implementations. We didn't want to be using `Box<dyn Error>` everywhere, meaning we wanted typed errors, so we utilized `thiserror`'s ability to generate `From` implementations from a macro to handle the transformation of our error types across the module and crate boundaries. (`thiserror` lets you succinctly define a `From` implementation like so: `IoError(#[from] std::io::Error)`).

Reviewer: ```suggestion
Up until recently our approach to error handling centered around using [`thiserror`](https://crates.io/crates/thiserror), a fantastic crate which abstracts away a lot of Rust's verbosity when creating errors. Previously we would create enumerations (generally large, top-level ones, per crate and/or major module) to describe our different error types, and use `thiserror`'s macros to handle a lot of the trait implementations. We didn't want to be using `Box<dyn Error>` everywhere, meaning we wanted typed errors, so we utilized `thiserror`'s ability to generate `From` implementations from a macro to handle the transformation of our error types across the module and crate boundaries. (`thiserror` lets you succinctly define a `From` implementation like so: `IoError(#[from] std::io::Error)`).
```

### Handling errors across module and crate boundaries

> This led us into a pattern where we'd have a function called across a boundary (say from the `engine` module to the `simulation` crate), necessitating that we transform it between error types, and so for ergonomic reasons, we'd create a `From` implementation for the whole `Error` enumeration:

Reviewer: ```suggestion
Our approach led us to adopt a pattern where we'd call functions across boundaries (for example, from the `engine` module to our `simulation` crate), necessitating transformation between error types. For ergonomic reasons, we'd then create a `From` implementation for the whole `Error` enumeration:
```

```rust
// mod engine

#[derive(ThisError)]
enum Error {
    #[error("Simulation error: {0}")]
    Simulation(#[from] crate::simulation::Error),
    ...
}
```

This would then often turn into error messages that looked like the following:

```plain
Error: Engine Error: Simulation Error: Experiment Error: Datastore Error: No such file or directory (os error 2)
```

These error messages would tell us very little about the context that led to the error and weren't very helpful in actually tracking down the causes of problems. And while we have logging infrastructure in place, which is helpful with identifying events that lead to errors, it relies on a few things:

- the logging level is set to something detailed enough to point the person debugging to the correct location (not knowing how to reproduce the error can make it difficult to just rerun with `RUST_LOG=trace`);
- there are actually enough log statements in the correct places to give enough information;
- the log statements have been written in a way that's helpful, and provide the correct information (which is a challenge by itself, as you'll often find you needed information you hadn't thought about before).

> Perhaps most importantly though, hEngine is a multi-threaded, multi-language project; this means that there will often be a mix of logs from various threads appearing before an error, with it being unclear as to which piece of logic caused the problem.

Reviewer: ```suggestion
Perhaps most importantly though, hEngine is a multi-threaded, multi-language project. The consequence of this is that there will often be a mix of logs from various threads appearing before an error, resulting in it not always being straightforward to tell which piece of logic caused a problem.
```

It's worth mentioning that `thiserror` is able to capture backtraces, which helps a fair bit. We started utilizing this by adding a `backtrace` field for some of our error variants, however, in doing so we ran into two issues.

Firstly, we needed to explicitly add the field to all error variants expected to have a backtrace:

```rust
// mod engine

#[derive(ThisError)]
enum Error {
    #[error("Simulation error: {0}")]
    Simulation(#[from] crate::simulation::Error, Backtrace),
    ...
}
```

Secondly, this required rerunning the program with `RUST_BACKTRACE=1`, or always running the program with it on.

> Additionally, while backtraces help us locate the line of code which generated the error (which helps with disambiguating the example error message shown above), it's still missing a lot of other information that could prove helpful, like which iteration of a loop caused the error, or what path was missing.

Reviewer: ```suggestion
Additionally, while backtraces help us locate the line of code which generated the error (which helps with disambiguating the example error message shown above), it's still missing a lot of other information that could prove helpful, such as which iteration of a loop caused the error, or what path was missing.
```

```rust
// mod experiment_controller
...
    .map_err(|experiment_err| Error::from(experiment_err.to_string()))?;
```

We already had an `experiment::Error` `From` `experiment_controller::Error` implementation, which meant that we were unable to define a `From` implementation going the other way. To get around this (initially temporarily, but as with many things, this ended up sticking around for a while) we simply turned the experiment error into a `String` and utilized the `Error::Unique` variant again.

This was clearly not ideal, and we ended up with a weird mixture of well-defined error variants and ones that were just captured in strings. The conversions also resulted in information loss depending on the `Display` implementation, including the loss of important pieces of information like the backtrace.

> ## So, why **another error handling library?**

Reviewer 2: This also looks a little strange on the rendered view, perhaps

```suggestion
## So, why **another** error handling library?
```

```plain
Could not parse configuration file
             at main.rs:6:10
      - Could not read file "config.json"
      - 1 additional opaque attachment

Caused by:
   0: No such file or directory (os error 2)
             at main.rs:5:10

Stack backtrace:
   0: error_stack::report::Report<T>::new
             at error-stack/src/report.rs:187:18
   1: error_stack::context::<impl core::convert::From<C> for error::report::Report<C>>::from
             at error-stack/src/context.rs:87:9
   2: <core::result::Result<T,E> as error::result::IntoReport>::report
             at error-stack/src/result.rs:204:31
   3: parse_config
             at main.rs:4:19
   4: main
             at main.rs:13:26
   5: ...

Suggestion: Use a file you can read next time!
```

### What's the catch?

You might think that this comes with a lot of performance overhead, but with speed being a key concern of hEngine, we've taken this into account. The impact on the ‘happy' success code path is minimal; a `Report` is only one pointer in size and requesting from a `Report` only happens at the time of error reporting. We use a similar data layout as `anyhow` and thus expect similar performance.

> However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that **more times than not** a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.

Reviewer 2: ```suggestion
However, we've purposefully added in some requirements that result in a little bit of extra upfront **development** overhead. We think having explicitly typed `Result` values is worth requiring, and especially helpful in identifying different types across module/crate boundaries, so we still want users to create types (`Context`s). This approach makes `error-stack` usable not only for binaries but also for libraries. In most cases, the caller of a library only cares about the error directly returned from a library, not an underlying OS error. As the current context is _always_ known, the user directly has all type information without optimistically trying to downcast to an error type (which remains possible). This also implies that more times than not a developer is _forced_ to add a new context because the type system requires it. We've found from experience that this automatically greatly improves error reporting.
```

The bold looks quite out of place here in the rendered view.

We haven't made it easy to create a new `Report` from a `String` because it requires the type to implement `Context`. It's still possible to use a newtype `UniqueError(String)` if a user wants to, but we discourage this usage because we've found that a lot of those places benefit from attaching a string message on a well-defined context instead. In any case, other libraries most probably still have to change the context from `UniqueError` to their context because of the explicit type requirement.

> 

Reviewer 2: This line break doesn't seem to be appearing in the rendered view :/

Reviewer: Resolving in https://github.com/hashintel/hash/pull/628/files#r894528626

> **By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.**

Reviewer: ```suggestion
By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.
```

> **By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.**

Reviewer 2: The bold on the rendered view is really intense, perhaps we don't put this whole thing in bold.

### Some other cool things about error-stack

> We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers. [Check out the repository](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) for even more information:

Reviewer: ```suggestion
We don't have space here to talk about all of the features of `error-stack`, but here are a few more teasers:
```

Will move the other sentence further down.

- The complete crate is written for `no-std` environments. However, when using `std`, a blanket implementation for `Context` for any `Error` is provided
- The blanket implementation for `Error` also makes the library compatible with almost all other libraries using the official trait
  - This means that if you pick this up, you should be able to migrate gradually (like we're doing with hEngine)
- Just like [`eyre`](https://crates.io/crates/eyre), the crate is able to set its own `Display` and `Debug` hooks that are called when printing the `Report`, which means you can write your own custom formatting and do other cool things
- We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` to capture a `SpanTrace`

> 

Reviewer: ```suggestion

[Check out the `error-stack` repository](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) for even more information about the library.

```

### The future of error-stack

As previously noted, we are still in the process of migrating our own codebase to `error-stack` and we intend to make further improvements as we apply the crate to more of hEngine and our internal code at [HASH](https://hash.ai/).

> During development of the library we included a `Provider` implementation while waiting for the implementation of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html) to be merged. We will be updating the library to swap to the `core::any` implementation in version 0.2.0 (This is the reason why **this API currently requires the nightly release channel**).

Reviewer: ```suggestion
During development of the library we included a `Provider` implementation while waiting for the implementation of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html) to be merged. This is the reason why the API currently requires the nightly release channel, and we will be updating the library to swap to the `core::any` implementation in version 0.2.0.
```

Up until recently our approach to error handling centered around using [`thiserror`](https://crates.io/crates/thiserror), a fantastic crate which abstracts away a lot of Rust's verbosity when creating errors. We used to create enumerations (generally a large, top-level one, per crate and/or major modules) to describe our different error types, and used `thiserror`'s macros to handle a lot of the trait implementations. We didn't want to be using `Box<dyn Error>` everywhere, meaning we wanted typed errors, so we utilized `thiserror`'s ability to generate `From` implementations from a macro to handle the transformation of our error types across the module and crate boundaries. (`thiserror` lets you succinctly define a `From` implementation like so: `IoError(#[from] std::io::Error)`).

> ### Handling Errors across Module and Crate Boundaries

Reviewer 2: ```suggestion
### Handling errors across module and crate boundaries
```

We have inconsistent capitalisation of our headers

Additionally, while backtraces help us locate the line of code which generated the error (which helps with disambiguating the example error message shown above), it's still missing a lot of other information that could prove helpful, like which iteration of a loop caused the error, or what path was missing.

> ### Defining Base Errors

Reviewer 2: ```suggestion
### Defining base errors
```

This was through no fault of `thiserror`, but our usage of it in practice, and the patterns we found ourselves settling into.

> ### Error Conversions

Reviewer 2: ```suggestion
### Error conversions
```

This was clearly not ideal, and we ended up with a weird mixture of well-defined error variants and ones that were just captured in strings. The conversions also resulted in information loss depending on the `Display` implementation, including the loss of important pieces of information like the backtrace.

> ## So, why **another Error Handling Library?**

Reviewer 2: ```suggestion
## So, why **another error handling library?**
```

- _Note: It's still possible to create something like `Error::Unique` but we do not provide shortcut functionality for that within the library, hoping that a lot of its uses can instead be solved through “attachments”._

> ## What is **`error-stack`?**

Reviewer 2: ```suggestion
## What is **error-stack?**
```
We're getting rid of code-formatted text in headings until we've managed to update the design to handle them better

`provide()` is an optional feature making use of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html?). When providing data here, they later can be requested by calling `Report::request_ref()` or `Report::request_value()`.

> **Note:** The implementation for the `Provider` RFC is not available on nightly at the time of writing, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any`. **Using the `Provider` API currently requires the Rust nightly compiler.**

Reviewer 2: ```suggestion
**Note:** The implementation for the `Provider` RFC is not available on nightly at the time of writing, so we have vendored in the functionality for now. We will switch to the upstream implementation in `core::any` with a future release of `error-stack`. **Using the `Provider` API currently requires the Rust nightly compiler.**
```

**By consciously imposing more restrictions in some areas we hope to reduce the _overall_ cost of development, in keeping with the philosophy underpinning Rust itself.**

> ### Some other cool things about `error-stack`

Reviewer 2: ```suggestion
### Some other cool things about error-stack
```
We're getting rid of code-formatted text in headings until we've managed to update the design to handle them better

- We have in-built support for backtraces on the nightly channel and feature-flagged support for `tracing` to capture a `SpanTrace`

### The future of `error-stack`

As previously noted, we are still in the process of migrating our own codebase to `error-stack` and we intend to make further improvements as we apply the crate to more of hEngine and our internal code at [HASH](https://hash.ai/).

During development of the library we included a `Provider` implementation while waiting for the implementation of the [`Provider` RFC](https://rust-lang.github.io/rfcs/3192-dyno.html) to be merged. We will be updating the library to swap to the `core::any` implementation in version 0.2.0 (This is the reason why this API currently **requires the nightly release channel**).

We cannot guarantee full backward compatibility until this crate hits `v1.0.0`. However, we will aim to avoid breaking changes, and in the event of any will announce them beforehand and provide a migration guide.

Contributions are greatly welcome. View the [project code on GitHub](https://github.com/hashintel/hash/tree/main/packages/libs/error-stack) and, if you like it, please star the repository!

> ### Acknowledgments

Reviewer 2: ```suggestion
## Acknowledgments
```

These aren't part of "What is error-stack" (I think I accidentally suggested you should change them)

We would like to thank the [Error Project Group](https://github.com/rust-lang/project-error-handling) for their work on the `Error` trait and the [Provider RFC](https://rust-lang.github.io/rfcs/3192-dyno.html). It's unlikely we would have written this library without the latter — or at least not in a way we would have been truly satisfied with.

> ### License

Reviewer 2: ```suggestion
## License
```
These aren't part of "What is error-stack" (I think I accidentally suggested you should change them)
